### PR TITLE
Forbid easily misused HashSet and HashMap constructors

### LINKED ISSUE #8423: Prohibit HashMap(capacity), HashMap(capacity, loadFactor), HashSet, LinkedHashMap constructors
They are pretty much always misused. See [this SO answer](https://stackoverflow.com/a/30220944/648955) for the explanation. They should be prohibited using forbidden-apis with suggested alternatives: Guava's `Maps.new(Linked)HashMapWithExpectedSize()`, `Sets.newHashSetWithExpectedSize()`.

### PR DESCRIPTION
**Pull request by capistrant**
Fixes #8423 

### Description

This PR adds constructors for HashMap, HashSet, and LinkedHashSet that specify an initial size parameter to the forbiddenapi list. It replaces all usage of these constructors with Guava helper methods that take the users parameter for size and create a more optimally sized object for them compared to what they get from the native constructors.

HashMap, HashSet and LinkedHashSet objects are often created in an inefficient way when the creator specifies a size. Guava has created utilities to take the size desired by the creator and come up with a proper initial size for the underlying object for them. It is explained more clearly [here](https://stackoverflow.com/a/30220944/648955). The same is also true for LinkedHashMap, but unfortunately we are still using Guava 16, and the Guava implementation for LinkedHashMaps did not come about until Guava 19.

I had to add an ignores item in the pom.xml file for `SomeAvroDatum.class` because it is generated at build time so it can't have a suppression annotation inline. If there is a better way to exclude that I'm not thinking of, please let me know.

<hr>

This PR has:
- [X] been self-reviewed
- [x] been tested in a test Druid cluster.

<hr>

##### Key changed/added classes in this PR

Many classes were modified, but the changes are all transparent. Just changing the way Map and Set objects are created in some specific cases.

### CODE REVIEW

**Comment by capistrant** [comment]
The travis-ci logs for the failed build steps all appear to be a 5XX error pulling external resources. Any way to get the build restarted short of pushing a dummy commit?

**Comment by jon-wei** [comment]
@capistrant I restarted the failed runs earlier, was going to comment but got pulled away for a moment

File: codestyle/druid-forbidden-apis.txt [hunk 1]
```diff
@@ -31,6 +31,12 @@ java.lang.String#replace(java.lang.CharSequence,java.lang.CharSequence) @ Use on
 java.lang.String#replaceAll(java.lang.String,java.lang.String) @ Use one of the appropriate methods in StringUtils instead, or compile and cache a Pattern explicitly
 java.lang.String#replaceFirst(java.lang.String,java.lang.String) @ Use String.indexOf() and substring methods, or compile and cache a Pattern explicitly
 java.nio.file.Files#createTempDirectory(java.lang.String prefix,java.nio.file.FileAttribute...) @ Use org.apache.druid.java.util.common.FileUtils.createTempDir()
+java.util.HashMap#<init>(int) @ Use com.google.common.collect.Maps#newHashMapWithExpectedSize(int) instead
+java.util.HashMap#<init>(int, float) @ Use com.google.common.collect.Maps#newHashMapWithExpectedSize(int) instead
+java.util.HashSet#<init>(int) @ Use com.google.collect.Sets#newHashSetWithExpectedSize(int) instead
+java.util.HashSet#<init>(int, float) @ Use com.google.collect.Sets#newHashSetWithExpectedSize(int) instead
+java.util.LinkedHashSet#<init>(int) @ Use com.google.collect.Sets#newLinkedHashSatWithExpectedSize(int) instead
+java.util.LinkedHashSet#<init>(int, float) @ Use com.google.collect.Sets#newLinkedHashSatWithExpectedSize(int) instead
```
Comment (leventov) [inline codestyle/druid-forbidden-apis.txt]: LinkedHashMap should be prohibited, too, even though there is no method ready in Guava. A polyfill method could be added to `CollectionUtils`.

File: core/src/main/java/org/apache/druid/utils/CollectionUtils.java [hunk 2]
```diff
@@ -39,6 +41,8 @@
 
 public final class CollectionUtils
 {
+  public static final int MAX_EXPECTED_SIZE = (1 << 30);
```
Comment (leventov) [inline core/src/main/java/org/apache/druid/utils/CollectionUtils.java]: Nit: should better be private.

Comment (capistrant) [reply core/src/main/java/org/apache/druid/utils/CollectionUtils.java]: Thanks for catching that

**Comment by capistrant** [comment]
Looks like a non-related TravisCI failure due to external dependency pulls

File: pom.xml [hunk 3]
```diff
@@ -1341,6 +1341,9 @@
                         <signaturesFile>${project.parent.basedir}/codestyle/joda-time-forbidden-apis.txt</signaturesFile>
                         <signaturesFile>${project.parent.basedir}/codestyle/druid-forbidden-apis.txt</signaturesFile>
                     </signaturesFiles>
+                    <excludes>
+                      <exclude>**/SomeAvroDatum.class</exclude>
```
Comment (leventov) [inline pom.xml]: I've created https://issues.apache.org/jira/browse/AVRO-2731 and https://github.com/apache/druid/issues/9331 to track this. FYI @Fokko.

@capistrant please try to create such issues yourself to close the loops and offload info from the heads of people.

**Pull request merged**